crates/amaru-kernel/src/memoized/transaction_output.rs

```diff
@@ -45,144 +50,102 @@ pub struct MemoizedTransactionOutput {
-                                    Err(e) => {
-                                        return Err(cbor::decode::Error::message(format!(
-                                            "failed to decode script: {:?}",
-                                            e
-                                        )));
-                                    }
-                                };
-                            }
-                            _ => {
-                                return Err(cbor::decode::Error::message(
-                                    "unknown tag for script tag",
-                                ));
-                            }
-                        }
-                    }
-                    _ => {
-                        return Err(cbor::decode::Error::message("unknown key option"));
-                    }
-                }
-            }
-        } else if data_type == cbor::data::Type::ArrayIndef || data_type == cbor::data::Type::Array
-        {
-            is_legacy = true;
-            d.array()?;
-            address =
-                Some(Address::from_bytes(d.bytes()?).map_err(|e| {
-                    cbor::decode::Error::message(format!("invalid address: {e:?}"))
-                })?);
-            value = d.decode_with(ctx)?;
-
-            if d.datatype()? == cbor::data::Type::Bytes {
-                datum = MemoizedDatum::Hash(d.bytes()?.into());
-            }
+        if data_type == Type::MapIndef || data_type == Type::Map {
+            decode_modern_output(d, ctx)
+        } else if data_type == Type::ArrayIndef || data_type == Type::Array {
+            decode_legacy_output(d, ctx)
         } else {
-            return Err(cbor::decode::Error::message(format!(
-                "unexpected CBOR type for output: {:?}",
-                data_type
-            )));
+            Err(cbor::decode::Error::type_mismatch(data_type))
         }
+    }
+}
 
-        let address = address
-            .ok_or_else(|| cbor::decode::Error::message("missing required address field"))?;
+fn decode_legacy_output<C>(
+    d: &mut cbor::Decoder<'_>,
+    ctx: &mut C,
+) -> Result<MemoizedTransactionOutput, cbor::decode::Error> {
+    let len = d.array()?;
+
+    Ok(MemoizedTransactionOutput {
+        is_legacy: true,
+        address: decode_address(d.bytes()?)?,
+        value: d.decode_with(ctx)?,
+        datum: match len {
+            Some(2) => MemoizedDatum::None,
+            Some(3) => {
+                let datum: Legacy<MemoizedDatum> = d.decode_with(ctx)?;
+                datum.0
+            }
+            Some(_) => {
+                return Err(cbor::decode::Error::message(format!(
+                    "expected legacy transaction output array length of 2 or 3, got {len:?}",
+                )))
+            }
+            None => {
+                let datum: Legacy<MemoizedDatum> = d.decode_with(ctx)?;
+                decode_break(d, len)?;
+                datum.0
+            }
```

Comment:
Will this work with an indefinite-length-encoded array that has two elements? it looks like this branch expects to find a datum.

Reply:
On closer inspection, the decode implementation for `Legacy<MemoizedDatum>` is looking for either a break (end of indefinite array) or a hash (though not consuming the break if it sees it). I think that's kind of confusing: it breaks an assumption I'd been making that `decode` only parses bytes belonging to the item getting decoded, and makes it look like we expect to consume bytes in this block when we actually don't.

More importantly, this is error-prone. `decode_break` won't error if the next element isn't a `break`, it'll just return `Ok(false)`. A 4-element indefinitely-encoded list will decode successfully here when it should've failed.

I'd recommend removing that break-checking logic from `Legacy<MemoizedDatum>` and moving it in here instead.
```suggestion
            None => {
                if decode_break(d, len)? {
                    MemoizedDatum::None
                } else {
                    let datum: Legacy<MemoizedDatum> = d.decode_with(ctx)?;
                    if !decode_break(d, len)? {
                        // return an error
                    }
                    datum.0
                }
            }
```